### PR TITLE
[fix] Detach KL teacher & fix per-layer weighting in AdaptiveLayerLoss

### LINKED ISSUE #3757: Question regarding KL Divergence gradient flow in `AdaptiveLayerLoss` (edit: and possible issue with layers weighting schema)
Hi @tomaarsen ,

I’ve been looking through the implementation of `AdaptiveLayerLoss` and had a question regarding the KL divergence computation between the intermediate layers and the final layer. I wanted to check if the current behavior is intended or if I might be misunderstanding a part of the implementation.

### Observation: 
In the `forward` method, the loss calculates `final_embeddings` from the full model to serve as a target for the intermediate layers:

```if self.kl_temperature > 0:    final_embeddings = forward_decorator.get_embeddings()    final_embeddings = F.softmax(final_embeddings / self.kl_temperature, dim=-1)```

Later, this is used as the `target` in `F.kl_div`:

```kl_div_loss = F.kl_div(   F.log_softmax(embeddings / self.kl_temperature, dim=-1),   final_embeddings,   reduction="batchmean",)```

(this seems to be common to both  `AdaptiveLayerLoss` and `Matryoshka2dLoss`)

### My Question: 
currently, `final_embeddings` is not detached from the computation graph. from my understanding, in PyTorch, this means that during the backward pass, gradients will flow not only into the intermediate `embeddings` (the "student", if we frame that as a distillation approach) but also back into the `final_embeddings` (the "teacher").

this creates a bidirectional optimization where the final layer is penalized for being different from the intermediate layers. In most distillation setups, ad far as i know, the teacher's output is typically detached, for obvious reasons. (still, one could argue if this really apply to this context of "self-distillation".

**Is it the intentional design of `AdaptiveLayerLoss` to allow the intermediate layers to influence the final layer's representations/optimization via this KL loss?**

or should the final layer be treated as a fixed reference point, like so?
```final_embeddings = F.softmax(final_embeddings / self.kl_temperature, dim=-1).detach()```

I'm curious to hear the thoughts of the maintainers on whether this bidirectional gradient flow is a feature of the 2DMSE approach or if it's something that might need a `.detach()`.

...a side (speculative) note:

interestingly, the recent 'Starbucks' paper (Zhuang et al., 2024, https://arxiv.org/pdf/2410.13230v1) notes that 2DMSE/ESE training can 'significantly hurt final-layer performance.' Given that they used the sentence-transformers implementation (*'For consistency, we implemented using the sentence transformers library which already included 2DMSE baseline implementations', section 4.2.1*), it's possible that the 'interference' they observed is actually the result of (or has some contribution from) this bidirectional gradient flow caused by the missing .detach() ?

btw, thanks for the great library!

Piero Boboli

*edit: fixed formatting/typo*
*edit: paper reference to v3: https://arxiv.org/pdf/2410.13230v3*

### PR DESCRIPTION
**Pull request by Kaif10**
### Fixes #3757

Two small, tightly-scoped fixes to `AdaptiveLayerLoss`, covering the points that were settled in the issue thread (thanks to @bobox2997 for the report/diagnosis and @wshuai190 and @tomaarsen for the discussion):

**1. Detach the KL self-distillation teacher.** `final_embeddings` (the final-layer output) is the target of the KL-divergence loss that aligns the prior layers with the last layer. It was not detached, so KL gradients flowed back into the final layer — pulling the teacher toward the intermediate students rather than the reverse. This adds `.detach()`, matching the standard stop-gradient teacher setup (agreed by @wshuai190 and @tomaarsen).

**2. Fix the per-layer weighting bracket typo.** The weighting divided by `(1 + layer_idx)` and then by `len(layer_indices)` separately, instead of by the intended bracketed fraction `((1 + layer_idx) / len(layer_indices))` that @tomaarsen identified as the intent in #3329. This only inserts the missing brackets so the existing intended formula is computed correctly (the two forms differ by a factor of `N²`).

**Explicitly out of scope:** I have *not* changed the default weighting *scheme* (linear vs. log vs. uniform). @wshuai190 noted the ESE/v2 paper uses log weighting and @tomaarsen indicated openness to updating the default — that's a separate design decision I've left to the maintainers.

**Tests.** Added two regression tests to `tests/sentence_transformer/losses/test_adaptive_layer.py`:
- the final transformer layer receives **zero** gradient from the KL term (teacher is detached), isolated by zeroing the non-KL weights;
- the total loss matches the intended `((1 + layer_idx) / len(layer_indices))` formula, not the un-bracketed one.

Both fail on `main` and pass with this change. `ruff check` / `ruff format` clean.

*Disclosure: prepared with the help of an AI coding assistant; I've reviewed and verified the change (repro of the gradient leak before/after and the weighting values) locally on CPU.*

### CODE REVIEW

**Comment by tomaarsen** [comment]
Hello!

Thanks for the tightly scoped PR, and thanks @bobox2997 and @wshuai190 for the diagnosis in #3757.

Re. the detach: fully agreed, and the regression test is a nice way to pin it down. I'm keeping both.

The weighting brackets: I went back to the sources before merging, and I have to walk back my own comment from #3329. The un-bracketed formula turns out to be faithful to the original works rather than a typo:

* The original 2DMSE reference implementation ([`angle.py` from Feb 2024](https://github.com/SeanLee97/AnglE/blob/347bbb6a9adf8cce293a41ebe462c5ca09b3812d/angle_emb/angle.py#L863)) computes `loss += student_loss / sample_layer` with one uniformly sampled layer per step. Our `layer_loss / (1 + layer_idx) / len(layer_indices)` is that same linear discount, generalized with an average for when multiple layers are sampled per step.
* The ESE paper ([arXiv 2402.14776v2](https://arxiv.org/abs/2402.14776v2)) uses `w_i = 1 / (1 + ln(i))` in Eq. 2 and states that "the weights decrease as the layer depth increases", with the last layer left unweighted. Its reference code also divides the student losses by `n_layers - 1`.
* The latest arXiv version ([v3](https://arxiv.org/abs/2402.14776v3)) reverts to the 2DMSE formulation: one uniformly sampled layer, no depth weighting at all.

So no version of the paper or the reference code ever gives the prior layers more weight than the final layer. The bracketed form would multiply the prior layer term by `len(layer_indices) ** 2`, which is a no-op with the default `n_layers_per_step=1` but roughly 144x for a 12 layer model with `n_layers_per_step=-1`. Apologies for pointing you at those brackets in the first place, that one is on me.

Since you allowed maintainer edits, I took the liberty of pushing a commit that drops the weighting change and its test, so this PR is now exactly the detach fix. I hope you don't mind. 

I'll open a separate PR that makes the weighting configurable via a `layer_weighting` parameter ("uniform", "log", "linear", or a custom callable), as proposed by @bobox2997 in the issue. In early experiments on STSb, "uniform" (which matches the latest 2DMSE version and the Starbucks paper) outperformed "log" and "linear" at every truncated layer count, so it is the planned new default.

Thanks again!

- Tom Aarsen
